### DISCUSSION TITLE
Can nanoFramework use the UI created by Gui-Guider

### DISCUSSION THREAD

**Discussion by TorchZhu** [nanoframework/Home#1329] [General]
Can nanoFramework integrate the LVGL code generated by Gui-Guider and how?

**Comment by TerryFogg** [comment]
LVGL is a great piece of code but has not been ported to nanoFramework. 
Porting is a very large amount of work.

**Reply by dotMorten** [reply]
You don't need to port it. LVGL is a C-API, so we just need p/invoke into it.

**Reply by alberk8** [reply]
I would like to see some work done there :)

**Reply by dotMorten** [reply]
I’m considering getting into nano framework on esp32 but need graphics heavy stuff so I might go down this rabbit hole. If so I’ll share it

**Reply by josesimoes** [reply]
@dotMorten that sounds like a nice improvement! 
Feel free to reach out before you commit time to this work. As this is something with potential impact on the existing UI stuff it will necessarily bring breaking changes. Just trying to avoid frustration and waste of work effort. 😉 

As for P/Invoke, we do have a similar technology with Interop, but is not _just_ calling the C API, there is a need to create wrappers around it to make it accessible from C#.

**Reply by Ellerbach** [reply]
Porting LVGL is **a lot** of work! Even if you automate code creation around a wrapper. And don't forget all the driver part, this should have to be adjusted as well. Last time I looked at this, I sized it as few months of full time work for a single developer. I know the framework is popular, that said, it's not perfect neither.

**Reply by jibie** [reply]
https://github.com/imxcstar/LVGLSharp

**Reply by Ellerbach** [reply]
This port is using generics, we don't have those yet. It also misses all the low level part. Again, not impossible, just a lot of work.

**Reply by josesimoes** [reply]
If I understood it correctly, that one is meant to "drive" LVGL from a Windows/Linux machine. So, quite out of scope for nano...

**Reply by alberk8** [reply]
> https://github.com/imxcstar/LVGLSharp

The P/Invoke in the used in project will only work on a full .Net 9,  A full porting to nanoframework is needed to get LVGL to work.
